### PR TITLE
Implement token authentication

### PR DESCRIPTION
**Pull request by dekimir**
A new kind of policy lets you forbid endpoint access without a particular HTTP header that contains the correct secret value.

### CODE REVIEW

**Comment by glommer** [comment]
This looks good to me in general, and tests are good.
First time I read the test I had the impression you were adding the token in plain text in the policy file. I then saw you aren't, but I wonder if I wasn't thrown off scent by it being pointed to by "value".

Maybe a more descriptive attribute name could help ?

**Comment by glommer** [comment]
ack, but tests fail

**Comment by dekimir** [comment]
@glommer tests fixed, added per-verb control.

File: cli/tests/integration_tests/framework.rs [hunk 1]
```diff
@@ -483,14 +487,15 @@ impl Chisel {
...
     /// Gets an `url` of a running ChielStrike service.
-    pub async fn get(&self, url: &str) -> Result<reqwest::Response> {
-        let url = url::Url::parse(&format!("http://{}", self.config.public_address))
-            .unwrap()
-            .join(url)
-            .unwrap();
+    pub async fn get(
+        &self,
+        url: &str,
+        headers: reqwest::header::HeaderMap,
```
Comment (honzasp) [inline cli/tests/integration_tests/framework.rs:493]: I think we should add a new method that supports request headers; most tests don't need to pass any headers, and this forces them to pass empty headers every time.

Comment (honzasp) [reply cli/tests/integration_tests/framework.rs:493]: (This applies to all other methods.)

File: cli/tests/integration_tests/rust_tests/test_token_auth.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
+use crate::framework::TestConfig;
+use crate::framework::{IntegrationTest, OpMode};
+use reqwest::{header::HeaderMap, StatusCode};
+use serde_json::json;
+
+fn header(name: &'static str, value: &str) -> HeaderMap {
+    let mut headers = HeaderMap::new();
+    headers.insert(name, value.parse().unwrap());
+    headers
+}
```
Comment (honzasp) [inline cli/tests/integration_tests/rust_tests/test_token_auth.rs:10]: We should perhaps consider adding these generic helper functions to `framework.rs`, so that they are available to all tests.

File: server/src/policies.rs [hunk 3]
```diff
@@ -71,10 +73,78 @@ impl UserAuthorization {
...
+}
+
+/// Describes a header that a request must include.
+#[derive(Clone, Default, Debug)]
+struct RequiredHeader {
+    header_name: String,
+    /// Names a secret (see secrets.rs) whose value must match the header value.
+    secret_name: String,
+    /// HTTP verbs to which this requirement applies.  If absent, apply to all verbs.
+    verbs: Option<Vec<hyper::Method>>,
```
Comment (honzasp) [inline server/src/policies.rs:140]: Maybe we should use "method" instead of "verb", to be consistent with the rest of the code?

**Review by honzasp — COMMENTED** [review]
This looks reasonable, but what are our intentions with this feature? Authentication based on a static global password is not going to be considered very secure in 2022 and it will not work at all when the backend is accessed from the client, so we probably should not market this as a canonical way to secure our backends.

File: cli/tests/integration_tests/rust_tests/test_token_auth.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+   mandatory_header: { name: header33, value: TOKEN33 }
+ - path: /
+   mandatory_header: { name: foo, value: BAR }"##,
+    );
+    chisel
+        .apply()
+        .await
+        .expect_err("Didn't catch repeat path")
+        .stderr()
+        .peek("Repeated path in header authorization");
```
[the remark this answers is not in the document]
Comment (dekimir) [reply cli/tests/integration_tests/rust_tests/test_token_auth.rs:81]: Isn't `read` slower?

File: server/src/policies.rs (hunk 3 above)
Comment (dekimir) [reply server/src/policies.rs:140]: "Method" is quite ambiguous, though.  We'd have to call it "HTTP methods" to disambiguate, which is inelegant.

File: cli/tests/integration_tests/rust_tests/test_token_auth.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply cli/tests/integration_tests/rust_tests/test_token_auth.rs:81]: I wasn't.  I read the source.  I can't imagine `.find` being faster than `.contains`.

**Comment by glommer** [comment]
> so we probably should not market this as a canonical way to secure our backends.

it doesn't have to be the canonical way, just a way, that is good enough for some simple applications.

These tokens can expire btw. You expire them by resetting your .env =p

File: cli/tests/integration_tests/framework.rs (hunk 1 above)
Comment (dekimir) [reply cli/tests/integration_tests/framework.rs]: Done.

File: cli/tests/integration_tests/rust_tests/test_array.rs [hunk 5]
```diff
@@ -60,15 +60,15 @@ pub async fn test_array(config: TestConfig) {
         .expect("failed to store data");
 
     {
-        let data = chisel.get_json("/dev/get").await;
+        let data = chisel.get_json("/dev/get", HeaderMap::new()).await;
```
[the remark this answers is not in the document]
Comment (dekimir) [reply cli/tests/integration_tests/rust_tests/test_array.rs]: Done 7d35de81

File: cli/tests/integration_tests/framework.rs [hunk 6]
```diff
@@ -547,6 +547,7 @@ fn chiseld() -> String {
 
 async fn setup_chiseld(config: &TestConfig) -> Result<(Database, GuardedChild, Chisel)> {
     let tmp_dir = Arc::new(TempDir::new("chiseld_test")?);
+    let path = tmp_dir.path().to_owned();
```
[the remark this answers is not in the document]
Comment (dekimir) [reply cli/tests/integration_tests/framework.rs]: Done 75523756

**Comment by dekimir** [comment]
Finally rebased. :sweat: All feedback has been addressed.

File: cli/tests/integration_tests/rust_tests/test_token_auth.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply cli/tests/integration_tests/rust_tests/test_token_auth.rs]: Let's not bikeshed this.  I can use `read()` as a default in the future.

File: server/src/policies.rs (hunk 3 above)
Comment (honzasp) [reply server/src/policies.rs:140]: It might be, but we use "method" everywhere else (both internally and in user-facing TypeScript APIs), so I guess it is even more inelegant to start using "verb" in policies.

Also, "method" is the correct technical term from the HTTP specifications, it is used almost universally in libraries, frameworks and APIs that deal with HTTP, and it is searched 10x more often than "verb" according to Google Trends [0].

[0] https://trends.google.com/trends/explore?date=all&q=http%20method,http%20verb
